File: scala/src/main/scala/RobotName.scala
import scala.annotation.tailrec
import scala.collection.mutable
import scala.util.Random

class Robot {
  var name = RobotName.next()

  def reset(): Unit = {
    name = RobotName.next()
  }
}

object RobotName {
  private val usedNames = mutable.HashSet.empty[String]

  private val letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
  private def randomLetter = letters(Random.nextInt(letters.length))
  
  private def randomName = f"${randomLetter}${randomLetter}${Random.nextInt(1000)}%03d"
  
  def next(): String = {
    val temp = randomName
    temp match {
      case name if usedNames contains name => 
        next()
      case name =>
        usedNames += name
        name
    }
  }
}

File: kotlin/src/main/kotlin/RobotName.kt
class Robot {
    companion object {
        private val NAMES = mutableSetOf<String>()
    }
    private var mutableName = generateName()

    private fun generateName(): String {
        var newName: String
        do {
            newName = "${randomLetter()}${randomLetter()}${randomDigit()}${randomDigit()}${randomDigit()}"
        } while (NAMES.contains(newName))
        NAMES.add(newName)
        return newName
    }

    private fun randomLetter() = ('A'..'Z').random()
    private fun randomDigit() = (0..9).random()

    val name: String
        get() = mutableName

    fun reset() {
        val originalName = mutableName
        mutableName = generateName()
        NAMES.remove(originalName)
    }
}
